lib/spack/spack/test/concretize_requirements.py

```diff
@@ -349,3 +349,51 @@ def test_default_and_package_specific_requirements(
     assert spec.satisfies(specific_exp)
     for s in spec.traverse(root=False):
         assert s.satisfies(generic_exp)
+
+
+@pytest.mark.parametrize("mpi_requirement", ["mpich", "mpich2", "zmpi"])
+def test_requirements_on_virtual(mpi_requirement, concretize_scope, mock_packages):
+    if spack.config.get("config:concretizer") == "original":
+        pytest.skip("Original concretizer does not support configuration" " requirements")
+    conf_str = """\
+packages:
+  mpi:
+    require: "{}"
+""".format(
+        mpi_requirement
+    )
+    update_packages_config(conf_str)
+
+    spec = Spec("callpath").concretized()
+    assert "mpi" in spec
+    assert mpi_requirement in spec
+
+
```

Comment:
(question) If the users asks for `hdf5 ^opempi` but we have

```
packages:
  mpi:
    require:
      mpich
```

will that work? Is it worth testing that this fails if it should be prevented?

Reply:
Done in 52d6620